datafusion/core/src/physical_optimizer/topk_aggregation.rs

```diff
@@ -0,0 +1,187 @@
+        if !kt.is_primitive() && kt != DataType::Utf8 {
+            return None; // TODO: other types?
+        }
+        if aggr.filter_expr.iter().any(|e| e.is_some()) {
+            return None;
+        }
+
+        // ensure the sort is on the same field as the aggregate output
+        let col = order.expr.as_any().downcast_ref::<Column>()?;
+        if col.name() != field.name() {
+            return None;
+        }
+
+        // We found what we want: clone, copy the limit down, and return modified node
+        let mut new_aggr = AggregateExec::try_new(
+            aggr.mode,
+            aggr.group_by.clone(),
+            aggr.aggr_expr.clone(),
+            aggr.filter_expr.clone(),
+            aggr.order_by_expr.clone(),
+            aggr.input.clone(),
+            aggr.input_schema.clone(),
+        )
+        .expect("Unable to copy Aggregate!");
+        new_aggr.limit = Some(limit);
+        Some(Arc::new(new_aggr))
+    }
+
+    fn transform_sort(plan: Arc<dyn ExecutionPlan>) -> Option<Arc<dyn ExecutionPlan>> {
+        let sort = plan.as_any().downcast_ref::<SortExec>()?;
+
+        // TODO: support sorting on multiple fields
+        let children = sort.children();
+        let child = match children.as_slice() {
+            [child] => child.clone(),
+            _ => return None,
+        };
+        let order = sort.output_ordering()?;
+        let order = match order {
+            [order] => order,
+            _ => return None,
+        };
+        let limit = sort.fetch()?;
+
+        let is_cardinality_preserving = |plan: Arc<dyn ExecutionPlan>| {
+            plan.as_any()
+                .downcast_ref::<CoalesceBatchesExec>()
+                .is_some()
+                || plan.as_any().downcast_ref::<RepartitionExec>().is_some()
+                || plan.as_any().downcast_ref::<FilterExec>().is_some()
+            // TODO: whitelist joins that don't increase row count?
+        };
+
+        let mut cardinality_preserved = true;
+        let mut closure = |plan: Arc<dyn ExecutionPlan>| {
+            if !cardinality_preserved {
+                return Ok(Transformed::No(plan));
+            }
+            if let Some(aggr) = plan.as_any().downcast_ref::<AggregateExec>() {
+                // either we run into an Aggregate and transform it
+                match Self::transform_agg(aggr, order, limit) {
+                    None => cardinality_preserved = false,
+                    Some(plan) => return Ok(Transformed::Yes(plan)),
+                }
+            } else {
+                // or we continue down whitelisted nodes of other types
+                if !is_cardinality_preserving(plan.clone()) {
+                    cardinality_preserved = false;
+                }
+            }
+            Ok(Transformed::No(plan))
+        };
+        let child = transform_down_mut(child, &mut closure).ok()?;
+        let sort = SortExec::new(sort.expr().to_vec(), child)
```

Comment:
Again, future work but seems like we can replace the final agg + sort with just the final agg (or even replace both with a `SortPreservingMergeExec(fetch = <limit>)`

Reply:
> tree nodes other than repartition/coalesce in between the partial and final agg?

I'd have to go through all the physical plan nodes to answer that for sure. Anything that doesn't alter row count should be fine, so projections, certain joins, etc.

> we can replace the final agg + sort

Ironically, the heap doesn't know anything about sorting, so something has to do that sort. We can do it in this PR @ `log2(limit)` by draining the heap, or quicksort can do it later at `limit * log2(limit)`. Either way I expect it to not be very expensive since `limit` should be small.

Same with partial and final aggregates. In Ballista I'd think we'd need to retain both, each worker can partially aggregate a million rows down to 10, then the final aggregate and merge 100 rows from 10 partitions into 10 rows total.

But in either case, since the bulk of the work has been pushed down to the leaves, inefficiencies higher up aren't a huge concern.

Reply:
> We can do it in this PR @ log2(limit) by draining the heap

I think this makes sense

Reply:
> But in either case, since the bulk of the work has been pushed down to the leaves, inefficiencies higher up aren't a huge concern.

Agreed. It's probably a minor optimization in most cases. In a large distributed query where we have 10k partitions doing a `SortPreservingMergeExec` (or a `TopK` when that is ready) might be meaningfully better than the final agg though.

Reply:
> I think this makes sense

Actually, I'd like to save this for another PR if that's alright. It's not as simple as it sounds, since we don't want to clone IDs for performance reasons and the only methods `RawTable` makes available to remove-and-take-ownership are things like `drain()` and `into_iter()` which AFAICT don't guarantee any kind of order.

So it's pretty easy to implement `BinaryHeap::pop()` to get out a reverse-ordered list, but not to the go and correlate those with the entries in the map.

Reply:
Sure, no worries. Let's just file a ticket to track right before/after merging so we don't forget

Reply:
I just implemented the drain idea.